### ISSUE TITLE
Array destructuring null results in PossiblyNullArrayAccess

### ISSUE THREAD

**Issue by trowski** [vimeo/psalm#3377]
Ran across this when working on the new stream API for Amp.

```php
$a = null;
[$value] = $a; // PHP does not issue any error and sets $value to null.
```
https://psalm.dev/r/bc0d03cd72
https://3v4l.org/RvKMt

We plan to use this pattern by fulfilling a continuation promise with a single-element array of the next streamed value or `null` if the stream has completed. e.g.

```php
while ([$value] = yield $stream->continue()) {
    // $value a value yielded from the async stream of values.
}
```

At the moment, this results in:

```
ERROR: PossiblyNullArrayAccess - lib/Stream/MapStream.php:31:18 - Cannot access array value on null variable  (see https://psalm.dev/079)
            if ([$value] = yield $this->stream->continue()) {
```

**Comment by muglug** [comment]
Yeah, it’s fine in your case but in the general case array access on `null` is not desired - so just suppress the issue

**Comment by trowski** [comment]
If it was only something we intended on using internally, yes, I would just suppress the issue. However, this was meant to be public API, so consumers of the stream interface would also see this issue.

PHP allows this without error, so therefore I'd say that Psalm should handle an array destructure from `null` as potentially setting the destructured variables to `null`, without error. In most cases, an unexpected `null` of a destructured variable would issue an error where the variable is used.

**Comment by kelunik** [comment]
@muglug Right, in the general case array access on `null` is not desired, as PHP also issues a notice in that case: https://3v4l.org/0cqmc. However, it's supported on `list` destructuring. It's really helpful to offer iteration with a single method.
